### PR TITLE
net processing: Add support for getcfilters

### PR DESCRIPTION
**Pull request by jnewbery**
Support `getcfilters` requests when `-peerblockfilters` is set.

Does not advertise compact filter support in version messages.

### CODE REVIEW

**Comment by jnewbery** [comment]
This is the 4th PR taken from #18876. Please see that PR for details of the full changes to support BIP 157.

File: src/net_processing.cpp [hunk 1]
```diff
@@ -2045,6 +2047,49 @@ static bool PrepareBlockFilterRequest(CNode* pfrom, const CChainParams& chain_pa
     return true;
 }
 
+/**
+ * Handle a cffilters request.
```
Comment (theStack) [inline src/net_processing.cpp:2051]:
```suggestion
 * Handle a cfilters request.
```

**Review by theStack — COMMENTED** [review]
Concept ACK

File: src/net_processing.cpp (hunk 1 above)
Comment (jnewbery) [reply src/net_processing.cpp]: Fixed. Thanks!

**Comment by jnewbery** [comment]
rebased on master and fixed failing test.

File: src/protocol.h [hunk 2]
```diff
@@ -225,6 +225,17 @@ extern const char* GETBLOCKTXN;
...
+ * getcfilters requests compact filters for a range of blocks.
+ * Only available with service bit NODE_COMPACT_FILTERS as described by
+ * BIP 157 & 158.
+ */
+extern const char* GETCFILTERS;
+/**
+ * cfilter is a response to a getcfilters request containing a single compact
+ * filter.
+ */
+extern const char *CFILTER;
```
Comment (MarcoFalke) [inline src/protocol.h:238]: style-nit in commit ee8499d294

```suggestion
extern const char* CFILTER;
```

**Review by maflcko — COMMENTED** [review]
ACK 1231f6b3ba 🔁

<details><summary>Show signature and timestamp</summary>

Signature:

```
-----BEGIN PGP SIGNED MESSAGE-----
Hash: SHA512

ACK 1231f6b3ba 🔁
-----BEGIN PGP SIGNATURE-----

iQGzBAEBCgAdFiEE+rVPoUahrI9sLGYTzit1aX5ppUgFAlwqrYAACgkQzit1aX5p
pUhYrQv+IxEs+9TDoo1Lq/ba4oUD2SnLQmHUQmfVTu/+MIjw/cwY9DoZIwg6/iGs
6WwTmf8NwlNjAe9AHnrgDm1BhakIZYaXv+ZgCL7BRqpFujyh2jwjsDjMXWppMvjv
DSt1NsrRlinXQsjR7WCJXvNt6axnvCeRsqLeua+ZhIyF5kVwTPapaNVbNo91Wjgw
fRmBwxhI39ZnZpjljKGOF9x7cB5y0ySCeDy2FQqRDpBEkdprUoSu5KloEjD9nNge
t4dtogwO6rDARHPx9rZOekp5MnAAUS+hdUeUFV3Yiw08VVbTuEOCAeAsIB326fQm
QORabhfya6K8guMuvCXyThmIi79GmAti93ybgwDREIWmsMJj88H7vHn8AM5adMR8
Zu839huCzBaeEUk4SbD5CK7j/2E3cGMLQ5l63m+4AEkOCgj/nr5aXqZpIGuxKU4F
ETGWpimT3U2VioNHBzoov/59uge5rPNoSpnKZukOaSYn00Uum1BFlgrpqDDjY1Pb
0D3QC8b5
=uTse
-----END PGP SIGNATURE-----
```

[Timestamp of file with hash `9f5d9f1dfb8cd50d63725419784f2b0521db1d74441352040b1817c8b21ec56f  -`](https://opentimestamps.org/info/?004f70656e54696d657374616d7073000050726f6f6600bf89e2e884e8929401089f5d9f1dfb8cd50d63725419784f2b0521db1d74441352040b1817c8b21ec56ff0106083ccf00d9a7fc42935f668636c4cf108fff0109b7bd67653ec3b9af95eb08e3eaea10608f120722543da6c31b6e73a068c4ac149a0ae2efb0fdc2ee503e6a20155c413060c1708f1045ecd41b0f0086175387c6626f7b20083dfe30d2ef90c8e2c2b68747470733a2f2f626f622e6274632e63616c656e6461722e6f70656e74696d657374616d70732e6f7267fff010d71cb0f614b37f86a38979cf7abf4da908f1045ecd41b0f008bcb4459d001d83010083dfe30d2ef90c8e232268747470733a2f2f6274632e63616c656e6461722e636174616c6c6178792e636f6dfff010ddee5ba7a1ee1d9566eba8f026c66f2108f1045ecd41b1f008f012f31bbea0e6b10083dfe30d2ef90c8e292868747470733a2f2f66696e6e65792e63616c656e6461722e657465726e69747977616c6c2e636f6df010ff46f6f5ef71d53a489287b5b149beba08f0203ee98f2b698420a0ee6758b2240a3f85e76e8342fd79201db417155e898e3e1108f1045ecd41b0f008f858a466f2daf1cf0083dfe30d2ef90c8e2e2d68747470733a2f2f616c6963652e6274632e63616c656e6461722e6f70656e74696d657374616d70732e6f7267)

</details>

File: src/protocol.h (hunk 2 above)
Comment (jnewbery) [reply src/protocol.h:238]: Fixed. Thanks!

File: src/net_processing.cpp [hunk 3]
```diff
@@ -3466,6 +3511,12 @@ bool ProcessMessage(CNode* pfrom, const std::string& msg_type, CDataStream& vRec
         return true;
     }
 
+    if (msg_type == NetMsgType::GETCFILTERS) {
+        ProcessGetCFilters(pfrom, vRecv, chainparams, connman);
+        return true;
+    }
+
+
```
Comment (Empact) [inline src/net_processing.cpp:3519]: nit: inconsistent whitespace

File: src/net_processing.cpp [hunk 4]
```diff
@@ -2051,6 +2053,49 @@ static bool PrepareBlockFilterRequest(CNode* pfrom, const CChainParams& chain_pa
...
+ * Handle a cfilters request.
+ *
+ * May disconnect from the peer in the case of a bad request.
+ *
+ * @param[in]   pfrom           The peer that we received the request from
+ * @param[in]   vRecv           The raw message received
+ * @param[in]   chain_params    Chain parameters
+ * @param[in]   connman         Pointer to the connection manager
+ */
+static void ProcessGetCFilters(CNode* pfrom, CDataStream& vRecv, const CChainParams& chain_params,
```
Comment (Empact) [inline src/net_processing.cpp:2066]: nit: since `pfrom` and `connman` are assumed, how about pass by reference?

**Review by Empact — COMMENTED** [review]
Code Review ACK https://github.com/bitcoin/bitcoin/pull/19044/commits/90a0dfff4ce4dec564f0a4b66504a68215aed27e

nit: for https://github.com/bitcoin/bitcoin/pull/19044/commits/0139a1b17cf6d44531b913443c77492e0881de72, how about noting in the commit message where this was introduced?

File: src/net_processing.cpp (hunk 3 above)
Comment (jnewbery) [reply src/net_processing.cpp:3519]: Fixed.

File: src/net_processing.cpp (hunk 4 above)
Comment (jnewbery) [reply src/net_processing.cpp:2066]: You're the second reviewer to ask for this so I've now made the change (as well as to `ProcessGetCFHeaders()` and `ProcessGetCFCheckPt`).

See also https://github.com/bitcoin/bitcoin/pull/19010#discussion_r429175250.

**Comment by jnewbery** [comment]
Thanks for the review @MarcoFalke and @Empact. I've addressed all of your review comments.

> nit: for 0139a1b, how about noting in the commit message where this was introduced?

The default [de]serialization wasn't previously used anywhere, so I don't think this is necessary. It's trivial to look up the git blame for that line.

File: src/net_processing.cpp (hunk 4 above)
Comment (Empact) [reply src/net_processing.cpp]: Thanks, IMO eliminating the null case is a worthwhile reason to break convention.

**Comment by Empact** [comment]
re-Code Review ACK https://github.com/bitcoin/bitcoin/pull/19044/commits/9e36067d8cd02830c7e5a88a391dff6ac3adbe0c

File: src/net_processing.cpp [hunk 5]
```diff
@@ -1998,7 +1998,7 @@ void static ProcessOrphanTx(CConnman* connman, CTxMemPool& mempool, std::set<uin
  * @param[out]  filter_index    The filter index, if the request can be serviced.
  * @return                      True if the request can be serviced.
  */
-static bool PrepareBlockFilterRequest(CNode* pfrom, const CChainParams& chain_params,
+static bool PrepareBlockFilterRequest(CNode& pfrom, const CChainParams& chain_params,
```
Comment (jkczyz) [inline src/net_processing.cpp:2003]: Does the "p" in `pfrom` stand for "pointer" or "peer"? Consider renaming if the former.

**Review by jkczyz — COMMENTED** [review]
ACK 9e36067d8cd02830c7e5a88a391dff6ac3adbe0c

File: src/net_processing.cpp [hunk 6]
```diff
@@ -2051,6 +2053,49 @@ static bool PrepareBlockFilterRequest(CNode& pfrom, const CChainParams& chain_pa
...
+    uint256 stop_hash;
+
+    vRecv >> filter_type_ser >> start_height >> stop_hash;
+
+    const BlockFilterType filter_type = static_cast<BlockFilterType>(filter_type_ser);
+
+    const CBlockIndex* stop_index;
+    BlockFilterIndex* filter_index;
+    if (!PrepareBlockFilterRequest(pfrom, chain_params, filter_type, start_height, stop_hash,
+                                   MAX_GETCFILTERS_SIZE, stop_index, filter_index)) {
```
[the remark this answers is not in the document]
Comment (jnewbery) [reply src/net_processing.cpp:2080]: I think the BIP and code are correct/consistent. If the difference between the start height and stop height in the request is 999, then the number of filters served is 1000. For example, request(1, 1000) results in blocks 1-1000 (1000 blocks) being served.

`if (stop_height - start_height >= max_height_diff)` ensures the strictly-less-than.

File: src/net_processing.cpp [hunk 7]
```diff
@@ -2051,6 +2053,49 @@ static bool PrepareBlockFilterRequest(CNode& pfrom, const CChainParams& chain_pa
...
+        return;
+    }
+
+    std::vector<BlockFilter> filters;
+
```
[the remark this answers is not in the document]
Comment (jnewbery) [reply src/net_processing.cpp:2085]: I'll do this if I need to retouch the branch.

File: src/net_processing.cpp (hunk 5 above)
Comment (jnewbery) [reply src/net_processing.cpp:2003]: I think you're right that the p stands for pointer. I'll rename these if I need to retouch the branch.

**Comment by MarcoFalke** [comment]
re-ACK 9e36067d8c , only change is adding commit "[refactor] Pass CNode and CConnman by reference" 🥑

<details><summary>Show signature and timestamp</summary>

Signature:

```
-----BEGIN PGP SIGNED MESSAGE-----
Hash: SHA512

re-ACK 9e36067d8c , only change is adding commit "[refactor] Pass CNode and CConnman by reference" 🥑
-----BEGIN PGP SIGNATURE-----

iQGzBAEBCgAdFiEE+rVPoUahrI9sLGYTzit1aX5ppUgFAlwqrYAACgkQzit1aX5p
pUgbKwv+PwO//7eg2MRGhEr5p5hlfuT6GXbIG2KBBj2at35/XQLwuNfi/0S8Jv8a
4oga+orbWYA4bNjZkEt2boyO3mCiML5pKPSHbjaDBaltE/lzzurIE0dfG/O+fPpk
+Smsj3HbC8BOXxkKdaqYXtQs5519LN9ypbCbIW6mY7SGyov1wG3XsBdNtMlU5Ybt
G7KkpTsXpSYpXCUy+Q1Rh2+3s7fvYZQrpe3XZ3Z0Ztzyuct3XgMarj+ucJOEKZ6Y
DKtXl0h4XBqfwlc4ngMZuXqN/6Zk1wEne94Q3Hr19SjbDLkp/oeBbsjqUi69y+5n
4CSmVaqhr8HUtTxqp/+ADSf+PaMchYx/jrOOh/f+2SsA0KLIsTvuGdWPxmDse0so
LHxFrDmF5mHXt1hECV5zQQcWE5M/nS8WB6nGHEQQ0RO67nxPs6CaqIQmR1WZ1599
/DAw2B1oagwy+KBQMAZFHjBFGKWa5TXXb7fZsVFSiUIXaMGl9p9xxfe2kJ1mzm5X
BaTr8GqR
=/Tog
-----END PGP SIGNATURE-----
```

[Timestamp of file with hash `e987c18ac02e885972b23102e7c1c01949ebe68ab8a6bbda08846816f39e846a  -`](https://opentimestamps.org/info/?004f70656e54696d657374616d7073000050726f6f6600bf89e2e884e892940108e987c18ac02e885972b23102e7c1c01949ebe68ab8a6bbda08846816f39e846af01050b14dbacf8c1d6e95990baf5a30682908fff0108591a4d7d48405d1615b69ff67377dd008f1045ed109c6f008aab0c3c9cce36fdf0083dfe30d2ef90c8e232268747470733a2f2f6274632e63616c656e6461722e636174616c6c6178792e636f6dfff010a0611ba46a38f3ddf3fb175b3d3f601508f1045ed109c6f008b0f29cf5848bdfea0083dfe30d2ef90c8e2c2b68747470733a2f2f626f622e6274632e63616c656e6461722e6f70656e74696d657374616d70732e6f7267fff010cee0ca55b6d5461a3cbf4c131b59e11e08f1045ed109c7f008a60d42cf5f1424ae0083dfe30d2ef90c8e292868747470733a2f2f66696e6e65792e63616c656e6461722e657465726e69747977616c6c2e636f6df010f17a05cc9322c8b7e54157bccc6634df08f1045ed109c7f0080c8e7ed007550f560083dfe30d2ef90c8e2e2d68747470733a2f2f616c6963652e6274632e63616c656e6461722e6f70656e74696d657374616d70732e6f7267)

</details>

File: src/net_processing.cpp (hunk 6 above)
[the remark this answers is not in the document]
Comment (jkczyz) [reply src/net_processing.cpp:2080]: Right, the difference. 👍

File: test/functional/p2p_blockfilters.py [hunk 8]
```diff
@@ -13,6 +13,7 @@
     hash256,
     msg_getcfcheckpt,
     msg_getcfheaders,
+    msg_getcfilters,
```
Comment (fjahr) [inline test/functional/p2p_blockfilters.py:16]: nit-ish: the comment up top in `p2p_blockfilters.py` is still missing `cfilters`.

**Review by fjahr — APPROVED** [review]
Code review ACK 9e36067d8cd02830c7e5a88a391dff6ac3adbe0c

If you have to retouch maybe also fix the commit message in 11106a4722558765a44ae45c7892724a73ce514c. It still says `-peercfilter` instead of `-peerblockfilter`.

File: test/functional/p2p_blockfilters.py (hunk 8 above)
Comment (MarcoFalke) [reply test/functional/p2p_blockfilters.py:16]: Let's fix this up in a follow-up

**Review by jonatack — COMMENTED** [review]
ACK 9e36067d8cd02830c7e5a88a391dff6ac3adbe0c

Two optional nits for follow-ups:
```diff
--- a/src/protocol.cpp
+++ b/src/protocol.cpp
@@ -40,12 +40,12 @@ const char *SENDCMPCT="sendcmpct";
-const char *GETCFILTERS="getcfilters";
-const char *CFILTER="cfilter";
-const char *GETCFHEADERS="getcfheaders";
-const char *CFHEADERS="cfheaders";
 const char *GETCFCHECKPT="getcfcheckpt";
 const char *CFCHECKPT="cfcheckpt";
+const char *GETCFHEADERS="getcfheaders";
+const char *CFHEADERS="cfheaders";
+const char *GETCFILTERS="getcfilters";
+const char *CFILTER="cfilter";

--- a/test/functional/p2p_blockfilters.py
+++ b/test/functional/p2p_blockfilters.py
@@ -4,8 +4,8 @@
 """Tests NODE_COMPACT_FILTERS (BIP 157/158).
 
-Tests that a node configured with -blockfilterindex and -peerblockfilters can serve
-cfheaders and cfcheckpts.
+Tests that a node configured with -blockfilterindex and -peerblockfilters can
+serve cfcheckpt, cfheaders, and cfilters.
 """
 ```

**Comment by jnewbery** [comment]
Thanks for the review @jonatack 

- the messages are defined in order [getcfilters, cfilter, getcfheaders, cfheaders, getcfcheckpt, cfcheckpt] in the BIP, so I think it makes sense to preserve that ordering in protocol.cpp.
- test docstring is fixed in the next PR

**Review by theStack — COMMENTED** [review]
post-merge Code Review ACK https://github.com/bitcoin/bitcoin/pull/19044/commits/9e36067d8cd02830c7e5a88a391dff6ac3adbe0c

**Pull request merged**